### PR TITLE
New feature #16832: Add user group functions to LimeSurveyAPI

### PR DESCRIPTION
**Pull request by flowdnb**
New feature #16832: Add user group functions to LimeSurveyAPI

Dev: Allow plugins to read and create user group, add/remove user from group

### CODE REVIEW

File: application/libraries/PluginManager/LimesurveyApi.php [hunk 1]
```diff
@@ -473,4 +473,90 @@ public function renderTwig($file, array $data)
...
+
+    /**
+     * Adds or removes a user from a user group
+     *
+     * @param integer $ugid The user group ID
+     * @param integer $uid The user ID
+     * @param string $action The desired action 'add' or 'remove'
+     * @return boolean True or false if user was added/removed to/from group or not
+     */
+    public function userInGroupAction($ugid, $uid, $action)
```
Comment (olleharstedt) [inline application/libraries/PluginManager/LimesurveyApi.php:528]: Please remove `$action` and split the function in two parts, one for remove, one for add. If the functions have code in common, split it into a third private function.

**Review by olleharstedt — CHANGES_REQUESTED** [review]

**Comment by Shnoulle** [comment]
Not about this function but a general remark : if we add function again on API we have a really great file finally  … maybe find another way to set "this part was in API" ?

File: application/libraries/PluginManager/LimesurveyApi.php [hunk 2]
```diff
@@ -473,4 +473,90 @@ public function renderTwig($file, array $data)
...
+    {
+        $group = \UserGroup::model()->findByAttributes(array('ugid' => $ugid));
+
+        if (empty($group)) {
+            return false;
```
Comment (Shnoulle) [inline application/libraries/PluginManager/LimesurveyApi.php:533]: Throw Exception ?

File: application/libraries/PluginManager/LimesurveyApi.php [hunk 3]
```diff
@@ -473,4 +473,90 @@ public function renderTwig($file, array $data)
...
+        } else {
+            if ($uid > 0 && \User::model()->findByPk($uid)) {
```
Comment (Shnoulle) [inline application/libraries/PluginManager/LimesurveyApi.php:535]: Same : throw Exception , no ?

File: application/libraries/PluginManager/LimesurveyApi.php [hunk 4]
```diff
@@ -473,4 +473,90 @@ public function renderTwig($file, array $data)
...
+     * @param string $groupDescription Description of user group to be created
+     * @return boolean True or false if user group was added or not
+     */
+    public function addUserGroup($groupName, $groupDescription)
+    {
+        $db_group_name = flattenText($groupName, false, true, 'UTF-8', true);
+        $db_group_description = $groupDescription;
+
+        if (isset($db_group_name) && strlen($db_group_name) > 0) {
+            if (strlen($db_group_name) > 21) {
```
Comment (Shnoulle) [inline application/libraries/PluginManager/LimesurveyApi.php:500]: No … must try to save or validate and return model->errors…

File: application/libraries/PluginManager/LimesurveyApi.php [hunk 5]
```diff
@@ -473,4 +473,90 @@ public function renderTwig($file, array $data)
...
+                return false;
+            } elseif (\UserGroup::model()->find("name=:groupName", array(':groupName'=>$db_group_name))) {
```
Comment (Shnoulle) [inline application/libraries/PluginManager/LimesurveyApi.php:502]: Same : try to validate : return model->errors

**Comment by olleharstedt** [comment]
Merged, but please consider also returning model errors if save fails.

**Comment by maziminke** [comment]
@olleharstedt can we also backport this to LS 3.x API? That version is the LTS version which most customers use so it makes sense making it available there as well.
It is just a new API function, that should not cause any harm...

**Comment by olleharstedt** [comment]
I can try to cherry-pick this into LS3 today. Shouldn't be a problem, as you say.

**Comment by olleharstedt** [comment]
Welp, forgot. Please remind me next week.

**Comment by olleharstedt** [comment]
Cherry-picked to LS3.

Reasons we approved internally:

* Zero or very low risk
* LS4 has been unstable long time
* Request by long-time partner

**Pull request merged**